### DISCUSSION TITLE
Problem Upgrade To Modoboa 1.15.0

### DISCUSSION THREAD

**Discussion by foxbat1989** [modoboa/modoboa#1964] [Upgrade]
# Impacted versions

* Modoboa: 1.15.0
* installer used: Yes
* Webserver: Nginx

# Steps to reproduce
Hello Guys
I Have Installed Modoboa version 1.14.0 on Ubunto 18.0.4 in Last Month
Now I Recived Update Notificatin in My WebMail Gui Panel
I Have Tried To upgrade By Instructions
When I use Command pip install modoboa==1.15.0, I See This Error :

Could not find a version that satisfies the requirement modoboa==1.15.0 (from versions: 0.7.0, 1.2.0rc2, 1.2.0, 1.2.1, 1.2.2, 1.3.0, 1.3.1, 1.3.2, 1.3.3, 1.3.4, 1.3.5, 1.4.1, 1.4.2, 1.4.3, 1.4.4, 1.4.5, 1.5.0, 1.5.1, 1.5.2, 1.5.3, 1.6.0, 1.6.1, 1.6.2, 1.6.3, 1.7.0, 1.7.1, 1.7.2, 1.7.3, 1.7.4, 1.8.0, 1.8.1, 1.8.2, 1.8.3, 1.9.0, 1.9.1, 1.10.0, 1.10.1, 1.10.2, 1.10.3, 1.10.4, 1.10.5, 1.10.6, 1.10.7, 1.11.0, 1.11.1, 1.12.0, 1.12.1, 1.12.2, 1.13.0, 1.13.1, 1.14.0)
No matching distribution found for modoboa==1.15.0

Please Help Me
Someone Please Give Me a Step By Step Upgrade Instruction

# Current behavior

# Expected behavior

# Video/Screenshot link (optional)

**Comment by Master13011** [comment]
tuto :
cd /srv/modoboa
mv env env.bck

git clone https://github.com/modoboa/modoboa-installer
cd modoboa-installer
sudo ./run.py --upgrade

**Comment by foxbat1989** [comment]
Hello 
i did Upgrade As your Instruction Commands
After Upgrade I Have Problem with Forwarding  email From My inbox
When i Select An Email From My Inbox And Click Forward, I see An Access Denied Error

**Comment by hadifarnoud** [comment]
I think we need a comprehensive tutorial on upgrading to 1.15.0 including rebuilding virtualenv and python3.

this is too risky for a beginner like myself so I probably won't do it. as of right now, we do have an issue loading modoboa dashboard but I don't want to risk upgrading

**Comment by ghost** [comment]
@Master13011 

sudo ./run.py --upgrade
usage: run.py [-h] [--debug] [--force] [--configfile CONFIGFILE]
              [--version {latest,1.9.0,1.8.1,1.8.2,1.8.3}]
              [--stop-after-configfile-check] [--interactive] [--upgrade]
              domain
run.py: error: too few arguments

**Comment by Master13011** [comment]
Édit :+1: 

sudo ./run.py --upgrade domain

**Comment by ghost** [comment]
This is what did NOT work for me:

cd /srv/modoboa
cp -R env env.bck
git clone https://github.com/modoboa/modoboa-installer
cd modoboa-installer
sudo ./run.py --version latest --upgrade --configfile installer.cfg  mail.domain.com

> Welcome to Modoboa installer!
> 
> Your mail server is about to be upgraded and the following components will be impacted:
> modoboa automx amavis clamav dovecot nginx razor postfix postwhite spamassassin uwsgi radicale opendkim
> Do you confirm? (Y/n) y
> The process can be long, feel free to take a coffee and come back later ;)
> Starting...
> Installing amavis
> Installing spamassassin
> Installing razor
> Installing clamav
> Installing modoboa
> User modoboa already exists, skipping creation but please make sure the /srv/modoboa directory exists.

sudo chown -R modoboa:modoboa /srv/modoboa

All seemed fine, ... and 

502 Bad Gateway
nginx

But I made a snapshot before I started, so I reverted, clearer documentation would help.

I'll try again later..

Amazing mail solution, hope better funding comes :)

**Comment by mirtouf** [comment]
In your virtual env, did you try _pip install modoboa==1.15_ ?

**Comment by ghost** [comment]
@mirtouf 

Lots os issues running the upgrade as the modoboa user inside the v env.
Got it going in a way, but have lots of warnings ..

`WARNINGS:
?: (modoboa-amavis.W001) AMAVIS_DEFAULT_DATABASE_ENCODING does not match the character encoding used by the Amavis database.
	HINT: Check your database character encoding and set/update AMAVIS_DEFAULT_DATABASE_ENCODING.
?: (security.W001) You do not have 'django.middleware.security.SecurityMiddleware' in your MIDDLEWARE so the SECURE_HSTS_SECONDS, SECURE_CONTENT_TYPE_NOSNIFF, SECURE_BROWSER_XSS_FILTER, and SECURE_SSL_REDIRECT settings will have no effect.`

My first run:

Sorry
An internal error occured.

Second run

Sorry
An internal error occured.

OMG, have to restore the image again, which means new ip's, all the records are going to change again, propogation ..

This should be easier.

**Comment by ghost** [comment]
Ok, I just did a new fresh install on the same server, some config, but that's fine.

**Comment by mirtouf** [comment]
In your settings.py, you should have something like this:
```
MIDDLEWARE = (
    'x_forwarded_for.middleware.XForwardedForMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'modoboa.core.middleware.LocalConfigMiddleware',
    'modoboa.lib.middleware.AjaxLoginRedirect',
    'modoboa.lib.middleware.CommonExceptionCatcher',
    'modoboa.lib.middleware.RequestCatcherMiddleware',
)
```

I cannot help about amavis as I am not using it.

**Comment by Toniob** [comment]
Are you serious @johnrb68 ? Do you think that's a way to talk ? Contributors and especially @tonioo are doing their best. If you have issues with Modoboa, maybe you should contribute a little, send patches, and stop whining. Or if you want your upgrade to be handled at best and you don't know how to do it, just pay somebody. Your message is unacceptable.
And this threat to get rid of the software… You know what ? Good riddance !

**Comment by tonioo** [comment]
@foxbat1989 Looks like a local pip cache issue. This has nothing to do with Modoboa, can you please try to upgrade your pip version and try again?

**Comment by mogwaai** [comment]
I have successfully upgraded from version 1.14 to 1.15 on Ubuntu 18.04 LTS. 

- Go to original root directory of Modoboa clone
- mv modoboa-installer/ modoboa-installer-1.14
- git clone https://github.com/modoboa/modoboa-installer
- cd modoboa-installer/
- cp ../modoboa-installer-1.14/installer.cfg .
- mv /srv/modoboa/env /srv/modoboa/env.bak
- ./run.py --upgrade mydomain.co.za

**Comment by hadifarnoud** [comment]
I have not used the installer. there are no tutorials on how to upgrade. may I ask why?

this is such a big change and confusing for so many, please do tell us how to upgrade if we're on Python 2 and not used the installer.

I want the bug fixes of this version too, thanks

**Comment by MrGeneration** [comment]
I'm confused.
The documentation states the steps required to upgrade.

Here, for example the steps for 1.15.0:
https://modoboa.readthedocs.io/en/latest/upgrade.html#id4

And, as you need to rebuilt your environment, this will help you on this step:
https://modoboa.readthedocs.io/en/latest/upgrade.html#rebuild-virtual-environment

Personally I don't think the docs should be more detailed on that part.
It tells you what to do, the python side of these steps is out of scope. (that's my opinion and view on things).

So what exactly is your issue?
Where are you stuck?

**Comment by hadifarnoud** [comment]
I don't know anything about Python. I think a documentation should be through regardless of user's knowledge.

**Comment by MrGeneration** [comment]
So you're saying the hint to this is not enough?
https://help.pythonanywhere.com/pages/RebuildingVirtualenvs/

You also disagree that Python is not application scope and thus shouldn't be seperate (not Modoboa related) documentation?
I really just want to understand your point of view as someone that's not a big fan of duplicate information all over the place.

**Comment by hadifarnoud** [comment]
yes I'm saying it has to be included in Modoboa docs. I want to make sure I'm doing it right, the sequence of steps etc.

since it's a production server and emails of hundreds of people are on it, I can't risk it and be down for hours (whilst we restore it from backup)

**Comment by tonioo** [comment]
@hadifarnoud Have you tried using the installer and its upgrade mode? Could you perhaps clone your server and try this?

**Comment by mikeosude** [comment]
Hi Guys;
I seem to be late to the party, I have a modoboa install 1.15.0 and I am trying to upgrade to the latest version 2.1.2 i have used git to clone the installer repo but when I try to run it there is no cfg file. I setup this server years ago when 1.15.0 was the latest version and i was always having issues when trying to upgrade so I leave it but i want to upgrade the server now and i really need help and proper step-by-step instructions to do the upgrade as this is a live production environment.

**Comment by cedr-x** [comment]
Still stuck on modoboa 1.14 / python 2.6...
Today, I tried my sunday to give a last try to rebuild modoboa virtual env with python3.6. 
A lot of requirements(.txt) does not work
pip install without version can fix some points, but MySQL-python is messing the rest.
Switching to a mature alternative like SOGo with a real effort on packaging is propably the best way to go
So long modoboa. 10 years.
